### ISSUE TITLE
Non-vacuum incidence medium

### ISSUE THREAD

**Issue by phoebe-p** [alexysong/inkstone#4]
Hi,

I have been working to integrate Inkstone into [RayFlare](https://github.com/qpv-research-group/rayflare), an open-source package developed in [our group at UNSW](https://www.qpvgroup.org) for simulating solar cells. I previously developed an interface with S4, which I have benchmarked extensively against e.g. the Fresnel equations and the transfer matrix method (for planar layers). So far, integrating Inkstone has been relatively straightforward because many of the methods are the same as S4, and for 'normal' cases (incidence medium = air/vacuum, n = 1), I am getting the same results with S4 and Inkstone. However, I cannot obtain reasonable results (i.e. in agreement with the Fresnel equations/TMM/S4) when the incidence medium has n > 1 (still non-absorbing). I attach two plots, for the following two stacks:

STACK 1:

n = 1 (incidence medium)
------
n = 4.12 + 0.33 i (50 nm)
------
n = 4.13 + 0.03 i (transmission medium)

STACK 2:

n = 1.52 (incidence medium)
------
n = 4.12 + 0.33 i (50 nm)
------
n = 4.13 + 0.03 i (transmission medium)

<img width="320" alt="image" src="https://github.com/alexysong/inkstone/assets/25822065/4ee6a643-86d0-4c79-9a86-0582350bf231">
<img width="320" alt="image" src="https://github.com/alexysong/inkstone/assets/25822065/a296fdbd-5f75-44a5-bf4f-c5de6f76abd7">

Calculated for 0 (normal) to 89.9 degree angle of incidence from the incidence medium (see note about incident polarisation below). The results for stack 1 are in excellent agreement at all angles (lines for TMM and Inkstone are on top of each other), while the results for stack 2 are not. I note that the result is correct at normal incidence, but not at other incidence angles, and I saw a comment in the code whi
[inkstone_inc_issue.txt](https://github.com/alexysong/inkstone/files/12886944/inkstone_inc_issue.txt)
ch implies perhaps this feature is simply not implemented yet: "todo: if incident not vacuum, then the incident wave shouldn't be Fourier order (which in general is not eigen)". Could you comment on whether there is currently a way to correct the result obtained for off-normal incidence, or if this is something that will be implemented in future?

Re: definition of incident polarisation, these seem to be conflicting between my definition in the TMM solver (which is the same as in S4) and that in Inkstone, but switching the s and p amplitudes gives the same results (for stack 1); these results are for Inkstone p-polarized incident light.

I have also attached the script which makes the plots, which requires the RayFlare package to work (`pip install rayflare`).

**Comment by alexysong** [comment]
Thank you for using inkstone. 

I think I know the reason behind this but need to confirm it. The example you attached requires `solcore` package for the material dispersion. Could you please attach it, or alternatively remove it? I think for this example we don't necessarily need dispersive material; you can use constant dielectric constants (which seems to be the case in your description but not in the code example.)

Polarization as explained in the front page:
$s$ polarization is when $\vec{E}$ field of the incoming plane wave is in the $xy$ plane. $p$ polarization is orthogonal to it.

**Comment by phoebe-p** [comment]
Hi Alex,
thanks for your very quick response! I have attached a new version of the file which doesn't use solcore or rayflare (since rayflare also requires solcore). I saved the TMM results calculated using rayflare as text files for comparison, also attached.
Phoebe
[TMM_air.txt](https://github.com/alexysong/inkstone/files/12897482/TMM_air.txt)
[TMM.txt](https://github.com/alexysong/inkstone/files/12897483/TMM.txt)
[inkstone_inc_issue_2.txt](https://github.com/alexysong/inkstone/files/12897488/inkstone_inc_issue_2.txt)

**Comment by alexysong** [comment]
A slightly longer explanation of is issue is as follows. This arises due to an awkward choice of setting the excitation due to a dilemma. Previously, the angles $\theta$ and $\phi$ were understood as if the incident was in the vacuum: if there was a vacuum before the infinite incident region; then in the actual solving there is no such vacuum before the infinite incident region. This awkward choice allows the incident region to be anything, including when patterned. See, in general, $s$ and $p$ are not even eigen in the incident region, and there is no canonical way of setting incident in these cases. 

Bottom line, the original computation makes sense when you understand $\theta$ and $\phi$ as if they were in a vacuum. Now this is fixed; the $\theta$ and $\phi$ are in the incident region material, which makes intuitive sense.

**Comment by phoebe-p** [comment]
Many thanks for resolving this so promptly!

**Comment by alexysong** [comment]
No worries. Please upgrade to 620708845269a8035db505ca28022335bdeebcfe. There was a bug causing $s$-polarization to be incorrect in uniform layers.

**Comment by alexysong** [comment]
Hi @phoebe-p,

Are you positive the TMM results are correct? Could you check this against S4? (I don't have S4 installed right now.)
Please don't upgrade to 0.3.12 yet.

Thanks,

**Comment by phoebe-p** [comment]
Yes, as I noted in my first post, I had to set the TMM polarisation to 's' and the Inkstone polarisation to 'p' to get the same result for incidence from air (and the TMM results match those of S4, with consistent definitions of the polarisation). So I think everything makes sense now, thanks!

**Comment by alexysong** [comment]
Great! Happy that this is resolved:)
